### PR TITLE
cmake: Convert attest logic to ctest

### PR DESCRIPTION
**Pull request by kwryankrattiger**
Update the test setup in cmake to use the CTest framework for filtering and running tests

### IBAMR Pull Request Checklist
- [ ] Does the test suite pass?
  - #1306 
- [ ] Was clang-format run?
- [x] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [ ] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [x] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [x] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

**Comment by kwryankrattiger** [comment]
Whoever reviews this, I have some questions on a couple things

- there is a test case name option `levels=<N>` but I couldn't see how that was used, so it is not in the new stuff. It seems like an old `mpirun=<N>` option
- Currently all of the test directories are added on the same level. Another option is to have a `CMakeLists.txt` under each diretory that adds the test drivers and the cases for those drivers.
- I set the default test timeout to 600 seconds, same as attest. I also saw 120s, but I wasn't sure if that should be the new default or if you only wanted that for CI.
- There are commented out messages for adding test directories/cases but it makes a ton of configure output. I am not sure if that would be useful or not.
- I do not "clean up" cases after they run for now. This could be added, but I think it was turned by default.

**Comment by drwells** [comment]
Nice work.

I am heading out of town until Monday so I will look at this next week!

**Comment by kwryankrattiger** [comment]
@drwells Hey just checking in to see if you have had a chance to review this.

**Comment by drwells** [comment]
Not yet - I only got back from camping last night. I'll get through this today or tomorrow.

**Comment by kwryankrattiger** [comment]
Another question I had for the tests  was on the connection between the examples and the tests. Are the test drivers (near) duplicates of the examples?

**Comment by drwells** [comment]
Only sometimes. I took a pretty big shortcut early on when writing tests of adding a lot of tests based on example programs, but more recently they are more like independent unit tests. Two notable examples:
1. IBFE/explicit/ex4 is my favorite example, so there are 37 tests based on the `tests/IBFE/explicit_ex4_2d` executable.
2. On the other hand `tests/interpolate/interpolate_01_2d` isn't based on any example and just checks all the IB kernels, and some in parallel (there are 24 tests for that executable).

**Comment by drwells** [comment]
I have answers to https://github.com/IBAMR/IBAMR/pull/1307#issuecomment-884406720:
- attest sets the test name equal to everything before the first period. There are a bunch of other pieces of information about tests that are ignored by attest after that point - for example, many interpolate tests include `.bspline_3` since they test the bspline 3 kernel. Unfortunately the wiki https://github.com/IBAMR/IBAMR/wiki/Tests-for-IBAMR is slightly out of date: we also support `expect_error=true` with the same syntax.
- Yes, we could split them, but the entire file is only 300 lines - adding 20 subfiles is overkill.
- I set the timeout on the CI pretty low since we had a lot of tests timing out and we wanted to fail fast. Now that we have pruned the list of running tests that can probably be removed - 600s is probably fine.
- I would prefer that we continue cleaning up tests since a lot of them generate huge amounts of output, and running in a new temporary directory guarantees that we are not reliant on untracked artifacts.

**Comment by drwells** [comment]
We are still missing 57 tests - attest can find 431 tests, but ctest only has 374.

One problem that I don't know how to fix with ctest is scheduling tests based on their CPU load. If I do `ctest -j8` right now I run 8 concurrent tests - this is not what I want, I want to use 8 concurrent cores.

**Review by drwells — CHANGES_REQUESTED** [review]
Another thing that isn't right with this is that tests are considered to pass if they successfully run  - i.e., a failing diff doesn't cause the whole test to fail.

File: cmake/IBAMRTesting.cmake [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+  if (NOT IBAMR_ENABLE_TESTING)
+    return ()
+  endif ()
+
+  #message(CHECK_START "Adding tests: ${test_DIRECTORY}/${name}")
+  #list(APPEND CMAKE_MESSAGE_INDENT "  ")
+
+  # Add test source if it exists (may not because of _2d/_3d suffix)
+  if (EXISTS "${IBAMR_TEST_DIR}/${test_DIRECTORY}/${name}.cpp")
+    list(APPEND test_SOURCES "${IBAMR_TEST_DIR}/${test_DIRECTORY}/${name}.cpp")
```
[the remark this answers is not in the document]
Comment (kwryankrattiger) [reply cmake/IBAMRTesting.cmake:86]: I am not sure what you mean my "automake-like names".

The "SOURCES" is there to make it easier for me to use the same CMake backend for 2d/3d/manual tests. In general, the sources should not required when adding a test and will report an error if the driver function is not detected.

File: cmake/IBAMRTesting.cmake [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+  # If there aren't any sources, the this is not going to work...
+  if (NOT test_SOURCES)
+    message(FATAL_ERROR "No sources detected for test: ${test_DIRECTORY}/${name}")
+  endif ()
+
+  # Find all of the test inputs
+  file(GLOB input_files "${IBAMR_TEST_DIR}/${test_DIRECTORY}/${name}.*.input")
+  if (NOT input_files)
+    #list(POP_BACK CMAKE_MESSAGE_INDENT)
+    #message(CHECK_FAIL "no test inputs detected, skipping...")
```
[the remark this answers is not in the document]
Comment (kwryankrattiger) [reply cmake/IBAMRTesting.cmake:98]: This was part of my question above related to writing out which tests are being added during configuration. I commented them out because they produced a lot of output but I wanted to get more guidance from you on what kind of output would be useful here, if any.

File: cmake/IBAMRTesting.cmake [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+    return ()
+  endif ()
+
+  ###############################################################################################
+  # Create build targets
+
+  # Add a top level test build target
+  if (NOT TARGET "ibamr-tests")
+    add_custom_target(ibamr-tests ALL)
```
[the remark this answers is not in the document]
Comment (kwryankrattiger) [reply cmake/IBAMRTesting.cmake:107]: So these targets are only added IFF `IBAMR_TESTING_ENABLED` is on. Adding it to `ALL` is just making sure they build if they are enabled.

File: cmake/IBAMRTesting.cmake [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+  # Connect the target to the directory target
+  add_dependencies("ibamr-tests-${test_DIRECTORY}" ${_target})
+
+  # Add the tests for this target
+  foreach (input ${input_files})
+    get_filename_component(input_stripped ${input} NAME_WLE)
+    get_filename_component(input_directory ${input} DIRECTORY)
+    set(check_output_file "${input_directory}/${input_stripped}.output")
+
+    _ibamr_parse_test(${input_stripped})
```
[the remark this answers is not in the document]
Comment (kwryankrattiger) [reply cmake/IBAMRTesting.cmake:139]: So, that is a good habit, and that is what the intention is. This function should only be called in this file, but CMake doesn't have a `static` equivalent that I know of. My solution, add an `_` in front of the function name that says it shouldn't be called elsewhere :)

File: cmake/IBAMRTesting.cmake (hunk 2 above)
[the remark this answers is not in the document]
Comment (drwells) [reply cmake/IBAMRTesting.cmake:98]: I'd prefer to just not output anything.

File: cmake/IBAMRTesting.cmake (hunk 3 above)
[the remark this answers is not in the document]
Comment (drwells) [reply cmake/IBAMRTesting.cmake:107]: True, but two things:
1. `IBAMR_TESTING_ENABLED` defaults to `ON`, but end users will typically not want to compile and run tests. Whatever way we go should be "tests are not compiled by default".
2. It's a style preference, but I'd prefer to keep `make tests` separate from just `make`.

File: cmake/IBAMRTesting.cmake (hunk 4 above)
[the remark this answers is not in the document]
Comment (drwells) [reply cmake/IBAMRTesting.cmake:139]: Got it. It's not that important (we can always homogenize the style later) but in that case the `_ibamr` is implied.

The other style habit I have is to use all capital letters for macros and functions (again, this is inherited from other projects I have worked on) so that all function calls look the same.

[the remark this answers is not in the document]
Comment (drwells) [reply cmake/IBAMRTesting.cmake:139]: If we don't export the macros then I suppose we don't need to preface them with `IBAMR_` - I don't think I was fully consistent in this when I wrote the rest of the build system.

File: cmake/IBAMRTesting.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (drwells) [reply cmake/IBAMRTesting.cmake:86]: by automake-like names I mean `test_SOURCES`, since that's the convention used with autotools' automake files:
https://github.com/IBAMR/IBAMR/blob/3922a88d91bbc9461aea61c76160455f4cfb6bbd/tests/IBFE/Makefile.am#L32

> In general, the sources should not required when adding a test and will report an error if the driver function is not detected.

Oh, I see - this is set up in a different way than I originally thought. I thought `test_SOURCES` was specific to exactly one executable.

File: cmake/IBAMRTesting.cmake [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+
+#[[!md
+
+ibamr_add_testss(
+  base_name
+  DIRECTORY
+    TestDirectory
+)
+
+]]#
```
[the remark this answers is not in the document]
Comment (kwryankrattiger) [reply cmake/IBAMRTesting.cmake:66]: This is the start of a doxygen comment for the CMake Test API, it is not commented code.

**Comment by kwryankrattiger** [comment]
@drwells

> Another thing that isn't right with this is that tests are considered to pass if they successfully run - i.e., a failing diff doesn't cause the whole test to fail.

Yes, that is true. Each test is actually 2 tests (3 if it is a restart test). The first test runs the executable and reports of the executable completed (`<test>_run`), the second test runs numdiff and reports if the test output matches the expected (`<test>_check`). It is possible to run both of these from the same script, but I always like breaking them apart because it makes it easier to determine what kind of error I am dealing with from the Failed tests log. I believe I can move all of this into a ctest script and run them all together as well, up to you.

> We are still missing 57 tests - attest can find 431 tests, but ctest only has 374.

That is strange, it was working briefly before. I don't include a test if it doesn't have an input file, attest would do that, but now that I am looking at this it also seems to not be running the check tests either, which isn't right. I will look at this more.

> One problem that I don't know how to fix with ctest is scheduling tests based on their CPU load. If I do `ctest -j8` right now I run 8 concurrent tests - this is not what I want, I want to use 8 concurrent cores.

Yes, I have some of the code there but commented it out. I forgot to circle back and add it back correctly to handle load issues.

**Comment by drwells** [comment]
Ah - splitting the tests is fine, but when I ran this on my machine I didn't see any of the `check` tests run (it sounds like you noticed the same issue).

File: cmake/IBAMRTesting.cmake (hunk 3 above)
[the remark this answers is not in the document]
Comment (kwryankrattiger) [reply cmake/IBAMRTesting.cmake:107]: Okay. I think the "CMake-y" way is to just not have tests enabled by default, and if you want to build and run tests then configure with the enable testing flag. CMake is very much configuration-centric.

File: cmake/IBAMRTesting.cmake (hunk 4 above)
[the remark this answers is not in the document]
Comment (kwryankrattiger) [reply cmake/IBAMRTesting.cmake:139]: This style is taken from Kitware projects. I wasn't sure what you would want to do since there weren't really any other CMake functions like this. I will go back and make all of the public CMake test API functions upper case.

**Comment by kwryankrattiger** [comment]
@drwells 
So I have done some digging and asking around and it seems the most reliable way to make sure all of the tests are added, and not skipped due to symlinks, is to list them explicitly in the add test function. If I were to do this, I would split these into subdirectories rather than list everything in the top level CMake.

```CMake
IBAMR_ADD_2D_TEST(explicit_ex1
  DIRECTORY IBEF
  CASES
    * # Looks for normal naming convention
    eliminate_pressure_jumps
    impose_pressure_jumps
    mpirun=4
    nodal_quadrature
    nodal_quadrature.mpirun=4
)
```

File: cmake/IBAMRTesting.cmake (hunk 5 above)
[the remark this answers is not in the document]
Comment (drwells) [reply cmake/IBAMRTesting.cmake:66]: How does this work?

**Comment by kwryankrattiger** [comment]
@drwells I re-pushed fixes for CPU balancing, fixed the logic around adding restart and check tests, and added a cleanup tests.

The issue with the symlinked tests is still there. I am just waiting on your thoughts on how to proceed.

New options (all marked advanced):
Option | Default | Usage
--- | --- | ---
`IBAMR_MAX_PARALLEL` | # System CPUs or 1 | Set the max CPU load/MPI ranks for tests
`IBAMR_TEST_CLEANUP_ON_FAILURE` |  ON | Cleanup test directories even if the test fails
`IBAMR_ALLOW_OVERSUBSCRIBE` | OFF | Allow adding tests if the CPUs required > `IBAMR_MAX_PARALLEL`
`IBAMR_TEST_TIMEOUT` | 600 | Test timeout
`IBAMR_TEST_DIR` | `${CMAKE_SOURCE_DIR}/tests/` | Root directory of tests

**Comment by drwells** [comment]
Yeah, I think we need to explicitly list the tests - it's tedious but better than globbing inside the build system.

This still has the problem with load balancing tests that need multiple processors:
```
drwells@vasavasorum build-homebrew % ctest -j8
Test project /Users/drwells/Documents/CPP/ibamr/build-homebrew
          Start  677: IBFE-explicit_ex4_2d.restart=50_run
          Start  634: IBFE-explicit_ex4_2d.inactive_1.mpirun=4.restart=50_run
          Start  568: CIB-cib_plate.restart=5.mpirun=4_run
          Start  581: IB-explicit_ex0.restart=300.mpirun=4_run
          Start  624: IBFE-explicit_ex4_2d.amr.restart=50.mpirun=4_run
          Start  681: IBFE-explicit_ex4_2d.restart=50.mpirun=5_run
          Start  706: IBFE-explicit_ex4_2d.scratch_hier.restart=25.mpirun=4_run
          Start  719: IBFE-explicit_ex5_2d.restart=40.mpirun=2_run
```
i.e., `ctest` will start jobs on 27 cores, even though I asked for 8.

**Comment by kwryankrattiger** [comment]
Another edge case that isn't behaving the way I would like is the `expect_failure=true` tests. If the case exits with an error code it is pretty easy to make these report as successes, but if the test aborts with an exception. I think it may be coming from the `TBOX_ERROR` macro, but I don't know what that does. Would you mind directing me to the project the provides that?

**Comment by drwells** [comment]
`TBOX_ERROR` comes from SAMRAI - it resolves down to calling `MPI_Abort()`, `abort()`, or `exit()`.

**Comment by drwells** [comment]
What's the issue with that case? If we abort then we have a nonzero exit code anyway.

File: cmake/IBAMRTesting.cmake [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+          list(APPEND _mpi_flags --oversubscribe)
+          set(load ${IBAMR_MAX_PARALLEL})
+        else ()
+          # Don't add tests that would require oversubscription
+          continue ()
+        endif ()
+      endif ()
+      list(APPEND _launcher
+        ${MPIEXEC} ${_mpi_flags} ${MPIEXEC_NUMPROC_FLAG} ${n_mpirun})
+      list(APPEND test_labels "MPI")
```
Comment (drwells) [inline cmake/IBAMRTesting.cmake:160]: I'm not sure where it should be added but we also need to disable multithreading with something like
```
export OMP_NUM_THREADS=1
export OMP_THREAD_LIMIT=1
```

Comment (kwryankrattiger) [reply cmake/IBAMRTesting.cmake:160]: Okay, that will be added down below where I setup properties.

**Comment by kwryankrattiger** [comment]
> What's the issue with that case? If we abort then we have a nonzero exit code anyway.

The issue seems to be the `WILL_FAIL` property on the test doesn't support `abort` codes, because those are escalated by CTest to a automatic failure since it is sending a signal, not just returning an non-zero error code. It isn't a big deal, I will just write a small cmake launcher for it to handle the special case.

**Comment by boyceg** [comment]
It also is possible to control what SAMRAI does in `TBOX_ERROR` in serial (although not in parallel).

**Comment by boyceg** [comment]
For a parallel job, is there something other than `MPI_Abort` that could be used to signal failure that would actually shut down MPI?

Ultimately, the shutdown happens in `SAMRAI_MPI::abort()` (defined in `source/toolbox/parallel/SAMRAI_MPI.C`).

**Comment by drwells** [comment]
The one test that utilizes this is `subdomain_level_translation_01.expect_error=true.input` - we could probably rig it up to exit(-1) instead.

**Comment by boyceg** [comment]
Any change that you are using `SAMRAI_MPI::setCallAbortInSerialInsteadOfExit()`? 😁

**Comment by drwells** [comment]
No - I don't think there is a way to toggle back to exit.

**Comment by kwryankrattiger** [comment]
I just noticed that this test was missing an input, but it has an output: `interpolate_velocity_02`

I am not sure how you want this case to be handled. Currently I am just skipping it.

**Comment by drwells** [comment]
Oh yeah, I remember that one:

https://github.com/IBAMR/IBAMR/blob/master/tests/IBFE/interpolate_velocity_02.cpp

We can probably delete that test :) Just ignore it for now.

**Comment by kwryankrattiger** [comment]
I am seeing this error. I am not sure if I am doing something wrong or not.

```
Output:
----------------------------------------------------------
'/build/ibamr/tests/IBTK/ibamr-test-IBTK-child_integrators_2d' '/cwd/tests/IBTK/child_integrators_2d.input'


Writing visualization files...


+++++++++++++++++++++++++++++++++++++++++++++++++++
At beginning of timestep # 0
Simulation time is 0
P=00000:Program abort called in file ``/cwd/ibtk/src/utilities/HierarchyIntegrator.cpp'' at line 673
P=00000:ERROR MESSAGE:
P=00000:Failed assertion: d_current_cycle_num + 1 == d_current_num_cycles
```

**Comment by kwryankrattiger** [comment]
@drwells

I re-pushed a pretty large refactor that _should_ resolve these issues:
* Adding all of the tests (I count 432 now!)
* `ctest -j 8 .` shouldn't use > 8 CPUs
* `expect_error=true` now reports `Passed` correctly
* correctly implement `IBAMR_TEST_CLEANUP_ON_FAILURE=FALSE` to not delete test directories when a test fails

I also removed `IBAMR_TEST_DIR` and the `ibamr_add_<2d_|3d_>test(... DIRECTORY <dir>)` option since these are now implicitly defined by the `CMakeLists.txt` that adds the test

Instead of requiring `"<default>"` for all of the tests that only have the one test, it is added by default if there are no other tests cases listed. It does not get added by default if there are other test cases listed though.

**Comment by drwells** [comment]
I don't know what's going on with that child integrator test. I'll take a look at your new version.

**Comment by abarret** [comment]
> I don't know what's going on with that child integrator test. I'll take a look at your new version.

I've also run into this showing up in an application. I think the issue is the dummy integrator in that test doesn't increment `d_current_cycle_num`. The error checking can be pretty fragile if the correct cycle numbers are not set correctly..

**Comment by drwells** [comment]
Yes, but it doesn't make any sense that the executable would abort when run under `ctest` and not under `attest.

**Comment by abarret** [comment]
> Yes, but it doesn't make any sense that the executable would abort when run under `ctest` and not under `attest.

It should only fail if `NDEBUG` is defined. I can reproduce this when running `attest`.

File: tests/IBTK/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+  LIBRARIES IBAMR2d
+  CASES
+    "<default>"
+    "mpirun=4"
+)
+ibamr_add_tests(mpi_type_wrappers
+  LIBRARIES IBAMR2d
+)
+
+ibamr_add_2d_tests(child_integrators)
```
Comment (abarret) [inline tests/IBTK/CMakeLists.txt:26]: This test has now been fixed on the main branch. If you rebase, this test is now dimension independent.

**Comment by kwryankrattiger** [comment]
@drwells I don't know if I have anything else to do on this, so if you guys are happy with how it works feel free to merge it.

**Comment by drwells** [comment]
I'll take a closer look at some point today or tomorrow.

**Comment by drwells** [comment]
I need another day!

**Comment by drwells** [comment]
I finally have some time to take a close look again. To start a bunch of tests are not set up correctly:

```
CMake Warning at cmake/IBAMRTesting.cmake:58 (message):
  Encountered unrecognized override levels
Call Stack (most recent call first):
  cmake/IBAMRTesting.cmake:100 (_ibamr_parse_test)
  cmake/IBAMRTesting.cmake:299 (ibamr_add_tests)
  tests/IBFE/CMakeLists.txt:12 (ibamr_add_2d_tests)


CMake Warning at cmake/IBAMRTesting.cmake:58 (message):
  Encountered unrecognized override levels
Call Stack (most recent call first):
  cmake/IBAMRTesting.cmake:100 (_ibamr_parse_test)
  cmake/IBAMRTesting.cmake:299 (ibamr_add_tests)
  tests/IBFE/CMakeLists.txt:12 (ibamr_add_2d_tests)


CMake Warning at cmake/IBAMRTesting.cmake:58 (message):
  Encountered unrecognized override levels
Call Stack (most recent call first):
  cmake/IBAMRTesting.cmake:160 (_ibamr_parse_test)
  cmake/IBAMRTesting.cmake:299 (ibamr_add_tests)
  tests/IBFE/CMakeLists.txt:12 (ibamr_add_2d_tests)


CMake Warning at cmake/IBAMRTesting.cmake:58 (message):
  Encountered unrecognized override levels
Call Stack (most recent call first):
  cmake/IBAMRTesting.cmake:160 (_ibamr_parse_test)
  cmake/IBAMRTesting.cmake:299 (ibamr_add_tests)
  tests/IBFE/CMakeLists.txt:12 (ibamr_add_2d_tests)


CMake Error at cmake/IBAMRTesting.cmake:115 (message):
  Cannot find input for test case: IBTK-child_integrators_2d-<default>
Call Stack (most recent call first):
  cmake/IBAMRTesting.cmake:299 (ibamr_add_tests)
  tests/IBTK/CMakeLists.txt:26 (ibamr_add_2d_tests)


-- Configuring incomplete, errors occurred!
```

I got around those by just commenting out the line creating the error.

A bunch of tests fail:

```
         29 - adv_diff-bp_adv_diff_02_2d.mpirun=8_run (Failed)
         30 - adv_diff-bp_adv_diff_02_2d.mpirun=8_check (Failed)
         41 - adv_diff-adv_diff_02_3d.mpirun=8_run (Failed)
         42 - adv_diff-adv_diff_02_3d.mpirun=8_check (Failed)
        445 - level_set-fe_surface_distance_2d.mpirun=8_run (Failed)
        446 - level_set-fe_surface_distance_2d.mpirun=8_check (Failed)
        453 - level_set-fe_surface_distance_3d.mpirun=8_run (Failed)
        454 - level_set-fe_surface_distance_3d.mpirun=8_check (Failed)
        465 - multiphase_flow-check_hydro_force.mpirun=8_run (Failed)
        466 - multiphase_flow-check_hydro_force.mpirun=8_check (Failed)
        841 - wave_tank-nwt_cylinder.mpirun=8_run (Failed)
        842 - wave_tank-nwt_cylinder.mpirun=8_check (Failed)
        845 - wave_tank-nwt.mpirun=8_run (Failed)
        846 - wave_tank-nwt.mpirun=8_check (Failed)
        1217 - IBFE-ib_partitioning_01_2d.tri3.b.mpirun=8_run (Failed)
        1218 - IBFE-ib_partitioning_01_2d.tri3.b.mpirun=8_check (Failed)
        1425 - IBTK-ibtk_mpi.mpirun=8_run (Failed)
        1426 - IBTK-ibtk_mpi.mpirun=8_check (Failed)
        1676 - IBTK-laplace_01_3d.mpirun=8_run (Failed)
        1677 - IBTK-laplace_01_3d.mpirun=8_check (Failed)
```

The load averages could be improved - I see
```
ctest -j8  4994.31s user 111.43s system 532% cpu 15:58.95 total
```
but `attest` finishes all tests in 14:34.54 with about 800% cpu.

1. Can we combine the setup and run tests and then also diff and cleanup?
2. I see that we have 1723 tests when we set it up this way - that's a prime number of tests. How did we get here if each test has a setup, run, diff, and cleanup subtest? I cannot figure out where this number is coming from since 432 * 4 = 1728.
3. there is a merge conflict - can you fix it?

**Comment by kwryankrattiger** [comment]
> I finally have some time to take a close look again. To start a bunch of tests are not set up correctly:

The warnings are there to notify that the name contains things that look like setup options, but are not doing anything. `level` is the main one, it looks like a legacy convention to `mpirun`.

The `<default>` tag issue should have been resolved....it seems I had some un-pushed fixes that I just sent that should at very least make the error more clear as to why it was failing.

> A bunch of tests fail:

I didn't see any of these...that is strange. I don't have Kitware time to spend on this, but I will try to take a look this weekend with some of my open source time.

> The load averages could be improved - I see

That is good, I am sure most of the overhead is scheduling the setup/run/check/cleanup as separate jobs.

>     1. Can we combine the setup and run tests and then also diff and cleanup?

Yes, this shouldn't be too hard now.

>     2. I see that we have 1723 tests when we set it up this way - that's a prime number of tests. How did we get here if each test has a setup, run, diff, and cleanup subtest? I cannot figure out where this number is coming from since 432 * 4 = 1728.

It is prime number is probably because some cases are missing output to check, so they have 3 instead of 4 steps. Funny that it ended up as a prime number, that is pretty arbitrary!

>     3. there is a merge conflict - can you fix it?

Yes, I will this weekend. This is a cool project, so I may do some other work on open issues. IB is something I want to learn more about! :smile:

**Comment by drwells** [comment]
Thanks a ton for continuing to work on this!

**Comment by drwells** [comment]
> level is the main one, it looks like a legacy convention to mpirun.

no, level here refers to levels in the AMR hierarchy - its just something to make the test names descriptive and isn't parsed by `attest`.

**Comment by kwryankrattiger** [comment]
@drwells Sorry that took so long to get to, the merge was non-trivial and I just got a little time this morning to get to it.

I am still working on merging the run test and compare, I am hoping to have it done this week!
